document-store/src/main/java/org/hypertrace/core/documentstore/postgres/query/v1/PostgresQueryParser.java

```diff
@@ -62,76 +62,68 @@ public String parse() {
 
     // handle where clause
     Optional<String> whereFilter = fromClause.isPresent() ? parseUnnestFilter() : parseFilter();
-    if (whereFilter.isPresent()) {
-      sqlBuilder.append(String.format(" WHERE %s", whereFilter.get()));
-    }
+    whereFilter.ifPresent(s -> sqlBuilder.append(String.format(" WHERE %s", s)));
 
     // selection clause
-    Optional<String> selectionClause = parseSelection();
-    if (selectionClause.isPresent()) {
-      sqlBuilder.insert(
-          startIndexOfSelection,
-          String.format("SELECT %s FROM %s", selectionClause.get(), finalTableName));
-    } else {
-      sqlBuilder.insert(startIndexOfSelection, String.format("SELECT * FROM %s", finalTableName));
-    }
+    sqlBuilder.insert(
+        startIndexOfSelection, String.format("SELECT %s FROM %s", getSelections(), finalTableName));
 
     // group by
     Optional<String> groupBy = parseGroupBy();
-    if (groupBy.isPresent()) {
-      sqlBuilder.append(String.format(" GROUP BY %s", groupBy.get()));
-    }
+    groupBy.ifPresent(s -> sqlBuilder.append(String.format(" GROUP BY %s", s)));
 
     // having
     Optional<String> having = parseHaving();
-    if (having.isPresent()) {
-      sqlBuilder.append(String.format(" HAVING %s", having.get()));
-    }
+    having.ifPresent(s -> sqlBuilder.append(String.format(" HAVING %s", s)));
 
     // order by
     Optional<String> orderBy = parseOrderBy();
-    if (orderBy.isPresent()) {
-      sqlBuilder.append(String.format(" ORDER BY %s", orderBy.get()));
-    }
+    orderBy.ifPresent(s -> sqlBuilder.append(String.format(" ORDER BY %s", s)));
 
     // offset and limit
     Optional<String> pagination = parsePagination();
-    if (pagination.isPresent()) {
-      sqlBuilder.append(String.format(" %s", pagination.get()));
-    }
+    pagination.ifPresent(s -> sqlBuilder.append(String.format(" %s", s)));
 
     return sqlBuilder.toString();
   }
 
-  private Optional<String> parseSelection() {
+  public String getSelections() {
+    return parseSelection().orElse(ALL_SELECTIONS);
+  }
+
+  public String getSelectionsWithImplicitId() {
+    return getSelections() + ", " + String.format("%s AS %s", ID, IMPLICIT_ID_ALIAS);
+  }
+
+  public Optional<String> parseSelection() {
     return Optional.ofNullable(PostgresSelectTypeExpressionVisitor.getSelections(this));
   }
 
-  private Optional<String> parseFilter() {
+  public Optional<String> parseFilter() {
```

Comment:
Currently, the parser is a stateful and parse thing in a specific order. So, not sure we should go it this way making independent clause methods public. Instead, we can add another parse method as per our need?

Reply:
Makes sense. Will merge `PostgresQueryBuilder` with this class.